### PR TITLE
[DynamoDB] Add waiter for table not exists after delete

### PR DESCRIPTION
**Pull request by GromNaN**
This waiter is in the official SDK.
https://github.com/aws/aws-sdk-php/blob/3.133.39/src/data/dynamodb/2012-08-10/waiters-2.json

Usage: 
```php
$dynamodbClient->deleteTable(['TableName' => 'errors']);
$dynamodbClient->tableNotExists(['TableName' => 'errors'])->wait();
```

### CODE REVIEW

File: src/Service/DynamoDb/src/Result/TableExistsWaiter.php [hunk 1]
```diff
@@ -11,18 +11,14 @@
 class TableExistsWaiter extends Waiter
 {
     protected const WAIT_TIMEOUT = 500.0;
-    protected const WAIT_DELAY = 20.0;
+    protected const WAIT_DELAY = 5.0;
```
Comment (GromNaN) [inline src/Service/DynamoDb/src/Result/TableExistsWaiter.php]: Whereas the default time is 20 seconds, the table are actually created in less that 5 seconds.

File: src/Service/DynamoDb/src/Result/TableNotExistsWaiter.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+use AsyncAws\Core\Exception\Http\HttpException;
+use AsyncAws\Core\Response;
+use AsyncAws\Core\Waiter;
+use AsyncAws\DynamoDb\DynamoDbClient;
+use AsyncAws\DynamoDb\Input\DescribeTableInput;
+
+class TableNotExistsWaiter extends Waiter
+{
+    protected const WAIT_TIMEOUT = 500.0;
+    protected const WAIT_DELAY = 2.0;
```
Comment (GromNaN) [inline src/Service/DynamoDb/src/Result/TableNotExistsWaiter.php]: Whereas the default time is 20 seconds, the empty table are actually deleted in less that 2 seconds.

File: src/Service/DynamoDb/src/Result/TableExistsWaiter.php [hunk 3]
```diff
@@ -11,18 +11,14 @@
...
 
     protected function extractState(Response $response, ?HttpException $exception): string
     {
         if (200 === $response->getStatusCode() && 'ACTIVE' === ($response->toArray()['Table']['TableStatus'] ?? null)) {
             return self::STATE_SUCCESS;
         }
 
-        if (null !== $exception) {
```
Comment (GromNaN) [inline src/Service/DynamoDb/src/Result/TableExistsWaiter.php]: This condition is the same as the following ternary condition.

File: src/Service/DynamoDb/src/Result/TableExistsWaiter.php (hunk 1 above)
Comment (jderusse) [reply src/Service/DynamoDb/src/Result/TableExistsWaiter.php]: this code is generated, you can't change it (CI fails)

this change have to be proposed in AWS reposity. Or we should think about a way to override it

File: src/Service/DynamoDb/src/Result/TableExistsWaiter.php (hunk 3 above)
Comment (jderusse) [reply src/Service/DynamoDb/src/Result/TableExistsWaiter.php]: please open an issue.

when the last check is about the exception, the ternary is not needed

**Comment by GromNaN** [comment]
The CI realigned me to the correct path. I'm updating the code generator & exception parser.

**Comment by Nyholm** [comment]
Make sure to update the generator in a second PR. That will make it easier to review and to merge quickly.

File: src/Core/src/Response.php [hunk 4]
```diff
@@ -87,18 +92,17 @@ public function __destruct()
...
-
-            if (400 <= $statusCode) {
-                throw $this->resolveResult = new ClientException($this->response);
-            }
-
-            if (300 <= $statusCode) {
-                throw $this->resolveResult = new RedirectionException($this->response);
-            }
+        if (500 <= $statusCode) {
+            $this->exception = new ServerException($this->response);
```
Comment (jderusse) [inline src/Core/src/Response.php]: Just realized that this code were wrong, calling this method several time with $throw=false then true does not return the rigth result. You don't need `this->exception`, we can use `this->resolveResult` to store the exception.

I think a best fix would be to remove the `$throw` parameter in this class,  and use try/catch in waiter to get the exception and silent errors

File: src/Core/src/Exception/Http/HttpExceptionTrait.php [hunk 5]
```diff
@@ -85,6 +85,9 @@ public function __construct(ResponseInterface $response)
                 $this->awsMessage = $body['Message'];
                 $message .= "\n\n" . $body['Message'] . "\n\n";
             }
+            if (isset($body['__type'])) {
+                list(, $this->awsCode) = explode('#', $body['__type'], 2);
```
Comment (jderusse) [inline src/Core/src/Exception/Http/HttpExceptionTrait.php]: SDK uses `$data['code'] = isset($parts[1]) ? $parts[1] : $parts[0];` meaning, it fallback to `parts[0]`

File: src/Service/Sqs/src/Result/QueueExistsWaiter.php [hunk 6]
```diff
@@ -19,7 +19,7 @@ protected function extractState(Response $response, ?HttpException $exception):
             return self::STATE_SUCCESS;
         }
 
-        if (null !== $exception && 'AWS.SimpleQueueService.NonExistentQueue' === $exception->getAwsCode() && 400 === $exception->getCode()) {
+        if (null !== $exception && 'QueueDoesNotExist' === $exception->getAwsCode() && 'AWS.SimpleQueueService.NonExistentQueue' === $exception->getAwsCode() && 400 === $exception->getCode()) {
```
Comment (GromNaN) [inline src/Service/Sqs/src/Result/QueueExistsWaiter.php]: Effet de bord.

File: src/CodeGenerator/src/Generator/WaiterGenerator.php [hunk 7]
```diff
@@ -214,17 +216,24 @@ private function getAcceptorPathBody(WaiterAcceptor $acceptor): string
 
     private function getAcceptorErrorBody(ErrorWaiterAcceptor $acceptor): string
     {
-        $checks = ['$exception !== null'];
+        $checks = [];
         $error = $acceptor->getError();
+        if (null !== $expected = $acceptor->getExpected()) {
```
Comment (jderusse) [inline src/CodeGenerator/src/Generator/WaiterGenerator.php]: IMHO this should be done only when the Exception does not have `error` attributes (The `QueueExistsWaiter` should not change)

Moreover, this is always true, as `getError()` called above is extract from `expected` value.

Do understand how it works, the `expected=ResourceNotFoundException` refer to a shape:
```yaml
    "ResourceNotFoundException":{
      "type":"structure",
      "members":{
        "message":{"shape":"ErrorMessage"}
      },
      "exception":true
    },
```
for SQS, this shape looks like
```
    "BatchEntryIdsNotDistinct":{
      "type":"structure",
      "members":{
      },
      "error":{
        "code":"AWS.SimpleQueueService.BatchEntryIdsNotDistinct",
        "httpStatusCode":400,
        "senderFault":true
      },
      "exception":true
    },
```

Is suggest to add the method `ExceptionShape::hasError`
```php
    public function hasError(): bool
    {
        return isset($this->data['error']);
    }
```

and change your code
```php
if (!$error->hasError()) {
    $checks[] = '$exception->getAwsCode() === ' . \var_export($expected, true);
} else {
    if (null !== $code = $error->getCode()) {
        $checks[] = '$exception->getAwsCode() === ' . \var_export($code, true);
    }
    if (null !== $code = $error->getStatusCode()) {
        $checks[] = '$exception->getCode() === ' . \var_export($code, true);
    }
}

File: src/Core/src/Response.php (hunk 4 above)
Comment (GromNaN) [reply src/Core/src/Response.php]: I've removed the `$throw` argument.

File: src/CodeGenerator/src/Generator/WaiterGenerator.php (hunk 7 above)
Comment (GromNaN) [reply src/CodeGenerator/src/Generator/WaiterGenerator.php]: Done.

File: src/Core/src/Exception/Http/HttpExceptionTrait.php (hunk 5 above)
Comment (GromNaN) [reply src/Core/src/Exception/Http/HttpExceptionTrait.php]: Done.

File: src/Core/src/Waiter.php [hunk 8]
```diff
@@ -120,13 +126,12 @@ public function __destruct()
...
      *
-     * @return bool whether the request is executed or not
-     *
      * @throws NetworkException
+     * @throws HttpException
      */
     final public function resolve(?float $timeout = null): bool
     {
-        return $this->response->resolve($timeout, false);
+        return $this->response->resolve($timeout);
```
Comment (jderusse) [inline src/Core/src/Waiter.php]: You should catch here

File: src/Core/src/Waiter.php [hunk 9]
```diff
@@ -64,8 +64,10 @@ public function __construct(Response $response, AbstractApi $awsClient, $request
 
     public function __destruct()
     {
-        if (null === $this->resolveResult) {
-            $this->resolve();
+        try {
+            $this->response->resolve();
```
Comment (jderusse) [inline src/Core/src/Waiter.php]: you can call `$this->resolve()` that should catch the exception

File: src/Core/src/Waiter.php [hunk 10]
```diff
@@ -163,8 +168,12 @@ public function __destruct()
 
         $start = \microtime(true);
         while (true) {
-            if (!$this->resolve($timeout - (\microtime(true) - $start))) {
-                break;
+            try {
+                if (!$this->response->resolve($timeout - (\microtime(true) - $start))) {
```
Comment (jderusse) [inline src/Core/src/Waiter.php]: use `$this->resolve()` that shoud catch the exception

File: src/CodeGenerator/src/Generator/WaiterGenerator.php [hunk 11]
```diff
@@ -214,17 +214,25 @@ private function getAcceptorPathBody(WaiterAcceptor $acceptor): string
...
     private function getAcceptorErrorBody(ErrorWaiterAcceptor $acceptor): string
     {
-        $checks = ['$exception !== null'];
+        $checks = [];
         $error = $acceptor->getError();
-        if (null !== $code = $error->getCode()) {
-            $checks[] = '$exception->getAwsCode() === ' . \var_export($code, true);
+        if ($error->hasError()) {
+            if (null !== $code = $error->getCode()) {
+                $checks[] = \var_export($code, true) . ' === $exception->getAwsCode()';
```
Comment (GromNaN) [inline src/CodeGenerator/src/Generator/WaiterGenerator.php]: I changed the order `value === variable` to align with what cs-fixer does.

**Comment by Nyholm** [comment]
Thank you for this PR.

**Pull request merged**